collect_app/src/main/java/org/odk/collect/android/widgets/PrinterWidget.kt

```diff
@@ -0,0 +1,134 @@
+package org.odk.collect.android.widgets
+
+import android.app.Activity
+import android.content.Context
+import android.graphics.Bitmap
+import android.util.Base64
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.ViewModel
+import com.google.android.material.button.MaterialButton
+import org.javarosa.core.model.data.IAnswerData
+import org.javarosa.form.api.FormEntryPrompt
+import org.jsoup.Jsoup
+import org.jsoup.nodes.Document
+import org.odk.collect.android.R
+import org.odk.collect.android.formentry.questions.QuestionDetails
+import org.odk.collect.android.utilities.QuestionMediaManager
+import org.odk.collect.androidshared.livedata.MutableNonNullLiveData
+import org.odk.collect.androidshared.livedata.NonNullLiveData
+import org.odk.collect.async.Scheduler
+import org.odk.collect.material.MaterialProgressDialogFragment
+import org.odk.collect.printer.HtmlPrinter
+import org.odk.collect.qrcode.QRCodeCreator
+import java.io.ByteArrayOutputStream
+
+class PrinterWidget(
+    context: Context,
+    questionDetails: QuestionDetails,
+    private val viewModel: PrinterWidgetViewModel,
+    private val scheduler: Scheduler,
+    private val questionMediaManager: QuestionMediaManager,
+    private val qrCodeCreator: QRCodeCreator,
+    private val htmlPrinter: HtmlPrinter
+) : QuestionWidget(context, questionDetails) {
+
+    init {
+        render()
+
+        MaterialProgressDialogFragment.showOn(
+            context as AppCompatActivity,
+            viewModel.isLoading,
+            context.supportFragmentManager
+        ) {
+            MaterialProgressDialogFragment().also { dialog ->
+                dialog.message = context.getString(org.odk.collect.strings.R.string.loading)
+            }
+        }
+    }
+
+    override fun onCreateAnswerView(context: Context, prompt: FormEntryPrompt, answerFontSize: Int): View {
+        val answerView = LayoutInflater.from(context).inflate(R.layout.printer_widget, null)
+        answerView
+            .findViewById<MaterialButton>(R.id.printer_button)
+            .setOnClickListener {
+                print()
+            }
+        return answerView
+    }
+
+    override fun setOnLongClickListener(listener: OnLongClickListener?) = Unit
+
+    override fun getAnswer(): IAnswerData? = null
+
+    override fun clearAnswer() = Unit
+
+    override fun registerToClearAnswerOnLongPress(activity: Activity?, viewGroup: ViewGroup?) = Unit
+
+    private fun print() {
+        formEntryPrompt.answerText?.let {
+            viewModel.parseAndPrint(scheduler, it, questionMediaManager, qrCodeCreator, context, htmlPrinter)
+        }
+    }
+}
+
+class PrinterWidgetViewModel : ViewModel() {
+    private val _isLoading = MutableNonNullLiveData(false)
+    val isLoading: NonNullLiveData<Boolean> = _isLoading
+
+    fun parseAndPrint(
+        scheduler: Scheduler,
```

Comment:
I think `Scheduler`, `QuestionMediaManager`, `QRCodeCreator` and `HtmlPrinter` could be constructor args for this class rather than method params.

Reply:
Could be but maybe not `QuestionMediaManager` since it's a ViewModel itself so passing one ViewModel as a parameter to another one would be weird. Done.

Reply:
Yeah I do agree that that feels a little icky. It's not actually a problem given that they both have the same "lifecycle", but still. This is probably a hint that the `QuestionMediaManager` shouldn't be implemented by a `ViewModel` and should instead be a component shared between them, but that's most likely a change for another time.